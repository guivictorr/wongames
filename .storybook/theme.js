import { create } from '@storybook/theming'
import theme from '../src/styles/theme'

export default create({
  base: 'dark',

  colorPrimary: theme.colors.primary,
  colorSecondary: theme.colors.secondary,

  // UI
  appBg: theme.colors.mainBg,
  appContentBg: theme.colors.mainBg,
  appBorderColor: theme.colors.lightGray,
  appBorderRadius: 4,

  // Typography
  fontBase: '"Open Sans", sans-serif',
  fontCode: 'monospace',

  // Text colors
  textColor: theme.colors.white,
  textInverseColor: theme.colors.primary,

  // Toolbar default and active colors
  barTextColor: theme.colors.white,
  barSelectedColor: theme.colors.primary,
  barBg: theme.colors.mainBg,

  // Form colors
  inputBg: theme.colors.lightGray,
  inputBorder: theme.colors.darkGray,
  inputTextColor: theme.colors.black,

  brandTitle: 'Won games storybook',
  brandUrl: 'https://example.com',
  brandImage: '/img/logo.svg'
})
