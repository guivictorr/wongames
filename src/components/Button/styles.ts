import styled, { css, DefaultTheme } from 'styled-components'
import { darken } from 'polished'
import { ButtonProps } from '.'

export type WrapperProps = {
  hasIcon?: boolean
} & Pick<ButtonProps, 'size' | 'fullWidth' | 'minimal'>

const buttonModifiers = {
  small: (theme: DefaultTheme) => css`
    height: 3rem;
    font-size: ${theme.font.sizes.xsmall};
  `,
  medium: (theme: DefaultTheme) => css`
    height: 4rem;
    font-size: ${theme.font.sizes.small};
    padding: ${theme.spacings.xxsmall} ${theme.spacings.medium};
  `,
  large: (theme: DefaultTheme) => css`
    height: 5rem;
    font-size: ${theme.font.sizes.medium};
    padding: ${theme.spacings.xxsmall} ${theme.spacings.xlarge};
  `,
  fullWidth: () => css`
    width: 100%;
  `,
  minimal: (theme: DefaultTheme) => css`
    background: none;
    color: ${theme.colors.primary};

    &:hover {
      color: ${darken(0.1, theme.colors.primary)};
    }
  `,
  withIcon: (theme: DefaultTheme) => css`
    svg {
      width: 1.5rem;
      & + span {
        margin-left: ${theme.spacings.xxsmall};
      }
    }
  `
}

export const Wrapper = styled.button<WrapperProps>`
  ${({
    theme,
    size = 'large',
    fullWidth = false,
    hasIcon,
    minimal = false
  }) => css`
    display: inline-flex;
    align-items: center;
    justify-content: center;
    color: ${theme.colors.white};
    text-decoration: none;
    background: linear-gradient(
      178.59deg,
      #ff5f5f -14.51%,
      #f062c0 102.86%,
      #f23131 102.86%
    );

    &:hover {
      background: ${minimal
        ? 'none'
        : 'linear-gradient(180deg, #e35565 0%, #d958a6 50%)'};
    }

    cursor: pointer;
    border-radius: ${theme.border.radius};
    border: 0;
    padding: ${theme.spacings.xxsmall};

    ${!!size && buttonModifiers[size](theme)}
    ${fullWidth && buttonModifiers.fullWidth()}
    ${!!hasIcon && buttonModifiers.withIcon(theme)}
    ${minimal && buttonModifiers.minimal(theme)}
  `}
`
