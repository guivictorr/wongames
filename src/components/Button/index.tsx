import * as S from './styles'

export type ButtonProps = {
  size?: 'small' | 'medium' | 'large'
  children?: React.ReactNode
  fullWidth?: boolean
  icon?: JSX.Element
  onClick?: (event: React.MouseEvent<HTMLButtonElement>) => void
}

const Button = ({ children, icon, ...props }: ButtonProps) => (
  <S.Wrapper hasIcon={!!icon} {...props}>
    {icon}
    {!!children && <span>{children}</span>}
  </S.Wrapper>
)

export default Button
