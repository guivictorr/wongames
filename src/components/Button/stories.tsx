import { Story, Meta } from '@storybook/react'
import { AddShoppingCart } from '@styled-icons/material-outlined/AddShoppingCart'
import Button, { ButtonProps } from '.'

export default {
  title: 'Button',
  component: Button,
  argTypes: {
    children: {
      type: 'string'
    },
    icon: {
      type: ''
    }
  }
} as Meta

export const Default: Story<ButtonProps> = (args) => <Button {...args} />
Default.args = {
  children: 'Buy now'
}
export const Icon: Story<ButtonProps> = (args) => <Button {...args} />
Icon.args = {
  children: 'Buy now',
  icon: <AddShoppingCart />
}
export const Link: Story<ButtonProps> = (args) => <Button {...args} />
Link.args = {
  children: 'Buy now',
  as: 'a',
  size: 'large',
  href: '/link'
}
