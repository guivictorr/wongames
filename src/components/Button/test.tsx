import { screen } from '@testing-library/react'
import { AddShoppingCart } from '@styled-icons/material-outlined/AddShoppingCart'
import { renderWithTheme } from 'utils/tests/helpers'

import theme from 'styles/theme'

import Button from '.'

describe('<Button />', () => {
  it('should render large button by default', () => {
    const { container } = renderWithTheme(<Button>Button</Button>)

    expect(screen.getByRole('button', { name: /Button/i })).toHaveStyle({
      height: '5rem',
      'font-size': theme.font.sizes.medium,
      padding: `${theme.spacings.xxsmall} ${theme.spacings.xlarge}`
    })

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render medium button when passed medium size', () => {
    renderWithTheme(<Button size="medium">Button</Button>)

    expect(screen.getByRole('button', { name: /Button/i })).toHaveStyle({
      height: '4rem',
      'font-size': theme.font.sizes.small,
      padding: `${theme.spacings.xxsmall} ${theme.spacings.medium}`
    })
  })

  it('should render small button when passed small size', () => {
    renderWithTheme(<Button size="small">Button</Button>)

    expect(screen.getByRole('button', { name: /Button/i })).toHaveStyle({
      height: '3rem',
      'font-size': theme.font.sizes.xsmall
    })
  })

  it('should render 100% width button when passed fullWidth size', () => {
    renderWithTheme(<Button fullWidth>Button</Button>)

    expect(screen.getByRole('button', { name: /Button/i })).toHaveStyle({
      width: '100%'
    })
  })

  it('should render button with icon when passed icon', () => {
    renderWithTheme(
      <Button icon={<AddShoppingCart data-testid="icon" />}>Button</Button>
    )

    expect(screen.getByText(/button/i)).toBeInTheDocument()
    expect(screen.getByTestId('icon')).toBeInTheDocument()
  })

  it('should render Button as a link', () => {
    renderWithTheme(
      <Button as="a" href="/link">
        Button
      </Button>
    )

    expect(screen.getByRole('link', { name: /button/i })).toHaveAttribute(
      'href',
      '/link'
    )
  })
})
