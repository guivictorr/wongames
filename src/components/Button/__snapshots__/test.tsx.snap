// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render large button by default 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  color: #FAFAFA;
  -webkit-text-decoration: none;
  text-decoration: none;
  background: linear-gradient( 178.59deg,#ff5f5f -14.51%,#f062c0 102.86%,#f23131 102.86% );
  cursor: pointer;
  border-radius: 0.4rem;
  border: 0;
  padding: 0.8rem;
  height: 5rem;
  font-size: 1.6rem;
  padding: 0.8rem 4.8rem;
}

.c0:hover {
  background: linear-gradient(180deg,#e35565 0%,#d958a6 50%);
}

<button
  class="c0"
>
  <span>
    Button
  </span>
</button>
`
